lib/Conversion/TorchOnnxToTorch/OnnxLstmExpander.cpp

```diff
@@ -0,0 +1,479 @@
+          // W shape [hidden_size, input_size]
+          Value Wb_i, Value Wb_o, Value Wb_f, Value Wb_c,
+          // Wb shape [hidden_size]
+          Value R_i, Value R_o, Value R_f, Value R_c,
+          // R shape [hidden_size, hidden_size]
+          Value Rb_i, Value Rb_o, Value Rb_f, Value Rb_c,
+          // Rb shape [hidden_size]
+          SmallVector<std::string> activations) {
+  Location loc = binder.getLoc();
+  auto intType = rewriter.getType<Torch::IntType>();
+  Torch::ValueTensorType HType =
+      H_prev.getType().cast<Torch::ValueTensorType>();
+
+  Value cstOne = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType, rewriter.getIntegerAttr(rewriter.getIntegerType(64), 1));
+
+  // Apply linear/matmul for each gate separately
+  // names are consistent with ONNX LSTM documentation
+  Value i_x = rewriter.create<Torch::AtenLinearOp>(loc, HType, Xt, W_i, Wb_i);
+  Value i_h =
+      rewriter.create<Torch::AtenLinearOp>(loc, HType, H_prev, R_i, Rb_i);
+  Value i =
+      rewriter.create<Torch::AtenAddTensorOp>(loc, HType, i_x, i_h, cstOne);
+  Value i_act = createActivationByName(rewriter, loc, activations[0], i);
+
+  Value o_x = rewriter.create<Torch::AtenLinearOp>(loc, HType, Xt, W_o, Wb_o);
+  Value o_h =
+      rewriter.create<Torch::AtenLinearOp>(loc, HType, H_prev, R_o, Rb_o);
+  Value o =
+      rewriter.create<Torch::AtenAddTensorOp>(loc, HType, o_x, o_h, cstOne);
+  Value o_act = createActivationByName(rewriter, loc, activations[0], o);
+
+  Value f_x = rewriter.create<Torch::AtenLinearOp>(loc, HType, Xt, W_f, Wb_f);
+  Value f_h =
+      rewriter.create<Torch::AtenLinearOp>(loc, HType, H_prev, R_f, Rb_f);
+  Value f =
+      rewriter.create<Torch::AtenAddTensorOp>(loc, HType, f_x, f_h, cstOne);
+  Value f_act = createActivationByName(rewriter, loc, activations[0], f);
+
+  Value ct_x = rewriter.create<Torch::AtenLinearOp>(loc, HType, Xt, W_c, Wb_c);
+  Value ct_h =
+      rewriter.create<Torch::AtenLinearOp>(loc, HType, H_prev, R_c, Rb_c);
+  Value ct =
+      rewriter.create<Torch::AtenAddTensorOp>(loc, HType, ct_x, ct_h, cstOne);
+  Value ct_act = createActivationByName(rewriter, loc, activations[1], ct);
+
+  Value C_forget =
+      rewriter.create<Torch::AtenMulTensorOp>(loc, HType, f_act, C_prev);
+  Value C_input =
+      rewriter.create<Torch::AtenMulTensorOp>(loc, HType, i_act, ct_act);
+  Value C_new = rewriter.create<Torch::AtenAddTensorOp>(loc, HType, C_forget,
+                                                        C_input, cstOne);
+
+  Value C_new_act =
+      createActivationByName(rewriter, loc, activations[2], C_new);
+  Value H_new =
+      rewriter.create<Torch::AtenMulTensorOp>(loc, HType, o_act, C_new_act);
+
+  return std::make_pair(H_new, C_new);
+}
+
+std::tuple<Value, Value, Value> lstm_layer( // returns Y, Y_h, Y_c
+    OpBinder binder, ConversionPatternRewriter &rewriter, Value X,
+    // X shape [seq_length, batch_size, input_size]
```

Comment:
Don't inline the shape information in the function declaration - it is better to just include a larger block comment up front with the information. You should also include some pseudo code or a link to the tensor math that represents an LSTM layer.